In [1]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv('../data/raw/german_credit_data.csv')
df

Unnamed: 0  Age     Sex  Job Housing Saving accounts Checking account  \
0             0   67    male    2     own             NaN           little   
1             1   22  female    2     own          little         moderate   
2             2   49    male    1     own          little              NaN   
3             3   45    male    2    free          little           little   
4             4   53    male    2    free          little           little   
..          ...  ...     ...  ...     ...             ...              ...   
995         995   31  female    1     own          little              NaN   
996         996   40    male    3     own          little           little   
997         997   38    male    2     own          little              NaN   
998         998   23    male    2    free          little           little   
999         999   27    male    2     own        moderate         moderate   

     Credit amount  Duration              Purpose  Risk  
0             1169         6             radio/TV  good  
1             5951        48             radio/TV   bad  
2             2096        12            education  good  
3             7882        42  furniture/equipment  good  
4             4870        24                  car   bad  
..             ...       ...                  ...   ...  
995           1736        12  furniture/equipment  good  
996           3857        30                  car  good  
997            804        12             radio/TV  good  
998           1845        45             radio/TV   bad  
999           4576        45                  car  good  

[1000 rows x 11 columns]

In [ ]:
#Handle nulls

df.isnull().sum()

Unnamed: 0            0
Age                   0
Sex                   0
Job                   0
Housing               0
Saving accounts     183
Checking account    394
Credit amount         0
Duration              0
Purpose               0
Risk                  0
dtype: int64

In [ ]:
#Since nulls appear on the accounts, we can assume that these individuals do not maintain either savings and/or checking accounts.
#We replace the Nulls with no account

df['Saving accounts'] = df['Saving accounts'].fillna('no account')
df['Checking account'] = df['Checking account'].fillna('no account')


In [ ]:
df.isnull().sum()

Unnamed: 0          0
Age                 0
Sex                 0
Job                 0
Housing             0
Saving accounts     0
Checking account    0
Credit amount       0
Duration            0
Purpose             0
Risk                0
dtype: int64

In [ ]:
df_decribed = df[['Sex', 'Age', 'Job', 'Credit amount', 'Duration']].groupby('Sex').describe()
df_decribed

Age                                                        Job  \
        count       mean        std   min   25%   50%   75%   max  count   
Sex                                                                        
female  310.0  32.803226  11.754363  19.0  24.0  29.0  37.0  75.0  310.0   
male    690.0  36.778261  10.988784  20.0  28.0  35.0  43.0  75.0  690.0   

                  ... Credit amount          Duration                       \
            mean  ...           75%      max    count      mean        std   
Sex               ...                                                        
female  1.835484  ...       3606.50  18424.0    310.0  19.43871  11.048047   
male    1.934783  ...       4266.25  15945.0    690.0  21.56087  12.437941   

                                     
        min   25%   50%   75%   max  
Sex                                  
female  4.0  12.0  18.0  24.0  60.0  
male    4.0  12.0  18.0  27.0  72.0  

[2 rows x 32 columns]

In [ ]:
# Distribution of housing by gender

pd.crosstab(df['Sex'], df['Housing'], normalize = 'index') * 100

Housing       free        own       rent
Sex                                     
female    6.129032  63.225806  30.645161
male     12.898551  74.927536  12.173913

In [ ]:
# Distribution of savings account by gender

pd.crosstab(df['Sex'], df['Saving accounts'], normalize = 'index') * 100

Saving accounts     little   moderate  no account  quite rich      rich
Sex                                                                    
female           62.580645  10.322581   15.806452    5.161290  6.129032
male             59.275362  10.289855   19.420290    6.811594  4.202899

In [ ]:
#Distribution of Checking account by gender

pd.crosstab(df['Sex'], df['Checking account'], normalize = 'index') *100

Checking account     little   moderate  no account      rich
Sex                                                         
female            28.387097  27.741935   37.419355  6.451613
male              26.956522  26.521739   40.289855  6.231884

In [ ]:
# Distribution of gender by loan purpose

pd.crosstab(df['Sex'], df['Purpose'], normalize = 'index') * 100

Purpose   business        car  domestic appliances  education  \
Sex                                                             
female    6.129032  30.322581             1.935484   7.741935   
male     11.304348  35.217391             0.869565   5.072464   

Purpose  furniture/equipment   radio/TV   repairs  vacation/others  
Sex                                                                 
female             23.870968  27.419355  1.612903         0.967742  
male               15.507246  28.260870  2.463768         1.304348

In [ ]:
#Proportion of risk profile per gender

pd.crosstab(df['Sex'], df['Risk'], normalize = 'index') * 100

Risk          bad       good
Sex                         
female  35.161290  64.838710
male    27.681159  72.318841

Descrptive insight from the data as extracted above:
- From the 'loan book' we have a large proportion of males to female (690:310)
- The proportion per gender that own the building they live in is considerably higher (63% of all female, and 74% of all male)
- The individuals in our loan book (to whom money has been lent out) seem to have low savings? Are they maintaining a savings account at another institution? Avegaes(little = 60%, moderate = 10%)
- The checking accounts look healthy (liitle = 27%, moderate = 27%); We are lending more to people with no checkings accounts?
- For the loan purpose by gender there is a fair distribution across all loan purpose apart from furniture/equipment. 8% more of women take loans for furniture and equipment.Biggest reason for a loan across all genders is (car > radio/tv > furniture/equipment).

Below we try and check the relationship between the variables (focus mostly on cat explanatory vars) against Risk.
Try and see how they relate.

In [ ]:
pd.crosstab(df['Housing'], df['Risk'], normalize = 'index') * 100

Risk           bad       good
Housing                      
free     40.740741  59.259259
own      26.086957  73.913043
rent     39.106145  60.893855

In [ ]:
pd.crosstab(df['Saving accounts'], df['Risk'], normalize = 'index') * 100

Risk                   bad       good
Saving accounts                      
little           35.986733  64.013267
moderate         33.009709  66.990291
no account       17.486339  82.513661
quite rich       17.460317  82.539683
rich             12.500000  87.500000

In [ ]:
pd.crosstab(df['Checking account'], df['Risk'], normalize = 'index') * 100

Risk                    bad       good
Checking account                      
little            49.270073  50.729927
moderate          39.033457  60.966543
no account        11.675127  88.324873
rich              22.222222  77.777778

In [2]:
df_test = pd.read_csv("../data/raw/german_credit_data.csv")
df_test

Unnamed: 0  Age     Sex  Job Housing Saving accounts Checking account  \
0             0   67    male    2     own             NaN           little   
1             1   22  female    2     own          little         moderate   
2             2   49    male    1     own          little              NaN   
3             3   45    male    2    free          little           little   
4             4   53    male    2    free          little           little   
..          ...  ...     ...  ...     ...             ...              ...   
995         995   31  female    1     own          little              NaN   
996         996   40    male    3     own          little           little   
997         997   38    male    2     own          little              NaN   
998         998   23    male    2    free          little           little   
999         999   27    male    2     own        moderate         moderate   

     Credit amount  Duration              Purpose  Risk  
0             1169         6             radio/TV  good  
1             5951        48             radio/TV   bad  
2             2096        12            education  good  
3             7882        42  furniture/equipment  good  
4             4870        24                  car   bad  
..             ...       ...                  ...   ...  
995           1736        12  furniture/equipment  good  
996           3857        30                  car  good  
997            804        12             radio/TV  good  
998           1845        45             radio/TV   bad  
999           4576        45                  car  good  

[1000 rows x 11 columns]

In [11]:
df_test.isna().sum()

df_test.columns

print(df_test.dtypes)


Unnamed: 0           int64
Age                  int64
Sex                 object
Job                  int64
Housing             object
Saving accounts     object
Checking account    object
Credit amount        int64
Duration             int64
Purpose             object
Risk                object
dtype: object
